webkit2/webkit2.uri-scheme-request.lisp

```diff
@@ -0,0 +1,95 @@
+(export 'webkit-uri-scheme-request-get-uri)
+
+(defcfun "webkit_uri_scheme_request_get_path" :string
+  (request webkit-uri-scheme-request))
+(export 'webkit-uri-scheme-request-get-path)
+
+(defcfun "webkit_uri_scheme_request_get_web-view" (g-object webkit-web-view)
+  (request webkit-uri-scheme-request))
+(export 'webkit-uri-scheme-request-get-web-view)
+
+(defcfun "webkit_uri_scheme_request_finish" :void
+  (request webkit-uri-scheme-request)
+  (stream :pointer)  ; XXX: GInputStream
+  (stream-length :int)
+  (contents :string))
+(export 'webkit-uri-scheme-request-finish)
+
+(defcfun ("webkit_uri_scheme_request_finish_error" %webkit-uri-scheme-request-finish-error) :void
+  (request webkit-uri-scheme-request)
+  (g-error :pointer))
+
+(defun webkit-uri-scheme-request-finish-error (request)
+  (glib:with-g-error (err)
+    (%webkit-uri-scheme-request-finish-error request err)))
+(export 'webkit-uri-scheme-request-finish-error)
+
+(cffi:defcallback uri-scheme-processed :void ((request :pointer) (user-data :pointer))
+  (let ((callback (find (cffi:pointer-address user-data) callbacks :key (function callback-id))))
+    (flet ((g-memory-input-stream-new-from-data (data-string)
+             (let* ((ffi-string (cffi:foreign-string-alloc data-string))
+                    ;; TODO: Can we rely on lisp's `length' here?
+                    ;; Will it give the same result as strlen on ffi-string?
+                    (ffi-string-length (length data-string))
+                    (stream (cffi:foreign-funcall "g_memory_input_stream_new_from_data"
+                                                  :string ffi-string
+                                                  :unsigned-int ffi-string-length
+                                                  ;; TODO: This should ideally be g_free()
+                                                  ;; to free the ffi-string automatically.
+                                                  :pointer (cffi:null-pointer))))
+               ;; ffi-string is returned because we need to free it afterwards
+               (values stream ffi-string ffi-string-length))))
+      (handler-case
+          (progn
+            (setf callbacks (delete callback callbacks))
+            ;; TODO: Use `unwind-protect' to free the allocated objects?
+            (when (callback-function callback)
+              (multiple-value-bind (data-type data)
+                  ;; Callback function should return data type (e.g., "text/html") as a first value
+                  ;; and data-string itself as a second value.
+                  (funcall (callback-function callback) request)
+                (multiple-value-bind (stream ffi-string data-length)
+                    (g-memory-input-stream-new-from-data data)
+                  (webkit-uri-scheme-request-finish request stream data-length data-type)
+                  (g:g-object-unref stream)
+                  ;; That's why g-memory-input-stream-new-from-data returns the string
```

Comment:
Why?  Typo?

Reply:
No, `g-memory-input-stream-new-from-data` returns the `ffi-string` for it to be freed later. Maybe it's not worth explaining twice, though.

Reply:
Oh, the comment was about the `g-object-unref` _above_, right?
I didn't get that.  To avoid this kind of ambiguity, I suggest writing comments _above_ the target code, ending with `:` to make it completely clear.

Reply:
No, it's about `cffi:foreign-string-free` below the comment :) 

What I meant was to emphasize that we need to free this string because it's created in `g-memory-input-stream-new-from-data` and cannot be freed there. This idea, however, should be understandable from a single comment inside the `g-memory-input-stream-new-from-data`. This comment we discuss is unnecessary and confusing. I'll remove it.